### PR TITLE
 OGR data provider is extemely slow on changing attribute values

### PR DESCRIPTION
**Pull request by landam**
See http://hub.qgis.org/issues/16216

### CODE REVIEW

File: src/providers/ogr/qgsogrprovider.cpp [hunk 1]
```diff
@@ -1534,6 +1534,12 @@ bool QgsOgrProvider::changeAttributeValues( const QgsChangedAttributesMap &attr_
 
   setRelevantFields( ogrLayer, true, attributeIndexes() );
 
+  bool hasTransaction = OGR_L_TestCapability( ogrLayer, OLCTransactions ) == TRUE;
+  if ( hasTransaction && OGR_L_StartTransaction( ogrLayer ) != OGRERR_NONE )
+  {
+    pushError( tr( "OGR error commiting transaction: %1" ).arg( CPLGetLastErrorMsg() ) );
```
Comment (m-kuhn) [inline src/providers/ogr/qgsogrprovider.cpp]: Should probably say "starting transaction"

File: src/providers/ogr/qgsogrprovider.cpp [hunk 2]
```diff
@@ -1534,6 +1534,12 @@ bool QgsOgrProvider::changeAttributeValues( const QgsChangedAttributesMap &attr_
 
   setRelevantFields( ogrLayer, true, attributeIndexes() );
 
+  bool hasTransaction = OGR_L_TestCapability( ogrLayer, OLCTransactions ) == TRUE;
```
Comment (m-kuhn) [inline src/providers/ogr/qgsogrprovider.cpp]: Can you either replace the `TRUE` with `true` or remove it (just for consistency)

**Comment by m-kuhn** [comment]
Thanks, looks good.

Any comment @rouault ?

File: src/providers/ogr/qgsogrprovider.cpp [hunk 3]
```diff
@@ -1647,6 +1653,11 @@ bool QgsOgrProvider::changeAttributeValues( const QgsChangedAttributesMap &attr_
     OGR_F_Destroy( of );
   }
 
+  if ( hasTransaction && OGR_L_CommitTransaction( ogrLayer ) != OGRERR_NONE )
+  {
+    pushError( tr( "OGR error commiting transaction: %1" ).arg( CPLGetLastErrorMsg() ) );
```
Comment (m-kuhn) [inline src/providers/ogr/qgsogrprovider.cpp]: typo: "committing"?

File: src/providers/ogr/qgsogrprovider.cpp (hunk 2 above)
Comment (landam) [inline src/providers/ogr/qgsogrprovider.cpp]: TRUE is defined by GDAL (cpl_port.h). It's safer to use TRUE I would say.

File: src/providers/ogr/qgsogrprovider.cpp (hunk 1 above)
Comment (landam) [inline src/providers/ogr/qgsogrprovider.cpp]: Right, fixed in 4edeb52

File: src/providers/ogr/qgsogrprovider.cpp (hunk 3 above)
Comment (landam) [inline src/providers/ogr/qgsogrprovider.cpp]: ops, fixed in e9380b0

File: src/providers/ogr/qgsogrprovider.cpp (hunk 2 above)
Comment (m-kuhn) [inline src/providers/ogr/qgsogrprovider.cpp]: Ok, works for me.
Code always looks a bit ancient to me when I see things like this :stuck_out_tongue:

**Comment by rouault** [comment]
LGTM. I'm wondering if that couldn't be generalized to doing StartTransaction in enterUpdateMode() and CommitTransaction in leaveUpdateMode() (except that we don't do an explicit call to leaveUpdateMode() in the provider destructor, so there would be a risk currently on non committed transaction). Anyway the PR as it is already a win.

**Comment by m-kuhn** [comment]
@rouault is there something preventing us from issuing `leaveUpdateMode()` in the destructor?

Btw, travis test result feedback is acting a bit strange right now. That's what it says:

```
======================================================================
FAIL: testDisablewalForSqlite3 (__main__.TestPyQgsOGRProviderGpkg)
Test disabling walForSqlite3 setting
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/qgis/QGIS/tests/src/python/test_provider_ogr_gpkg.py", line 393, in testDisablewalForSqlite3
    self.assertIsNone(cbk.msg)
AssertionError: 'OGR error starting transaction: Transaction already established' is not None
----------------------------------------------------------------------
Ran 9 tests in 0.696s
```

**Comment by rouault** [comment]
> is there something preventing us from issuing leaveUpdateMode() in the destructor?

None that I can't think to.

> travis test result feedback is acting a bit strange right now. That's what it says:

"Transaction already established" comes from https://github.com/OSGeo/gdal/blob/trunk/gdal/ogr/ogrsf_frmts/sqlite/ogrsqlitedatasource.cpp#L2802 . It could potentially mean that the explicit StartTransaction() call in the QGIS OGR provider comes whereas an implicit transaction is in progress by the GPKG driver itself, but nothing comes to mind why that would happen. Would require more careful debugging and breakpoints in OGRSQLiteBaseDataSource::SoftStartTransaction and OGRSQLiteBaseDataSource::StartTransaction

**Comment by gioman** [comment]
@landam I tried the patch and I confirm the very good speed improvements! thanks.

**Comment by m-kuhn** [comment]
@rouault what would you recommend to proceed with this?

Some I can think of

 * Be more permissive in the gpkg test (treat is as an expected warning with no bad side-effects)
 * Disable this for gpkg
 * Needs fixes in gdal and should be `#ifdef`'d with up-to-current gdal versions
 * ...

**Comment by rouault** [comment]
> what would you recommend to proceed with this?

What is the reason for the failed test ?

**Comment by m-kuhn** [comment]
see your comment above. the gpkg test chokes on the warning "Transaction already established"

**Comment by rouault** [comment]
ah indeed. already forgot it. well, reading again my comment, someone will have to debug more to understand what is really going on

**Comment by m-kuhn** [comment]
@landam does that sound like something you could do?
@rouault can you take the lead on this?

I don't know any details, but it sounds like a good thing to have merged.

**Comment by gioman** [comment]
> I don't know any details, but it sounds like a good thing to have merged.

it is indeed. On fairly large vectors I have seen improvements of at least 3x in the update operation.

**Comment by rouault** [comment]
@m-kuhn Do you know which GDAL revision (I see it is a 2.2.0dev) is used on Travis ? I can't reproduce locally the error in the gpkg test with latest GDAL trunk. Anyway we can be robust to a failed attempt at starting a transaction and silencing it.

**Comment by rouault** [comment]
Closing this PR. Updated it as PR #4257

**Pull request closed without merging**